relay-server/src/services/project/state.rs

```diff
@@ -0,0 +1,59 @@
+//! Types that represent the current project state.
+use std::sync::Arc;
+
+mod fetch_state;
+mod info;
+
+pub use fetch_state::{ExpiryState, ProjectFetchState};
+pub use info::{LimitedProjectInfo, ProjectInfo};
+use serde::{Deserialize, Serialize};
+
+/// Representation of a project's current state.
+#[derive(Clone, Debug)]
+pub enum ProjectState {
+    /// A valid project that is not disabled.
+    Enabled(Arc<ProjectInfo>),
+    /// A project that was marked as "gone" by the upstream. This variant does not expose
+    /// any other project information.
+    Disabled,
+    /// A project to which one of the following conditions apply:
+    /// - The project has not yet been fetched.
+    /// - The upstream returned "pending" for this project (see [`crate::services::project_upstream`]).
+    /// - The upstream returned an unparsable project so we have to try again.
+    /// - The project has expired and must be treated as "has not been fetched".
+    Pending,
```

Comment:
I'm a bit hesitant to call it "missing" because when a projectconfig response does not contain a key, we treat the project as disabled and we previously called this state "missing", see https://github.com/getsentry/relay/blob/c9bd1f809b33823fb225cd5e5fbe9d250e97d5bb/relay-server/src/services/project.rs#L122-L127